### DISCUSSION TITLE
Security Best Practices for Forms

### DISCUSSION THREAD

**Discussion by lollococce** [Budibase/budibase#4229] [Help] [answered]
Hi!

I have a Postgres SQL Table ('hr_applications') having some columns that carry sensitive information about the HR application review and some other columns such as applicant_name, applicant_email etc that are normally needed for a Create New Application Form. I usually connect the table directly under the Form Schema option but using access levels above Basic; in this case I need to gather applicants' info from a Public Screen and I wanna be extra careful. I feel like to be sure not to give any scope to those sensitive columns of the hr_applications table in the Create New Application Form I would need to use a SQL SELECT Query limiting the columns to those really necessary for the public user to be added for the field group instead of using the Form Schema option table and delete some fields from the field group - it feels like it's not the safest way or is it?

I might, in this case, split the table in postgres in a 'hr_applications' and 'hr_applications_metadata' (with the sensitive cols only) for example to be extra careful about security and go directly with the selection of the schema table, but that would complicate the design with a one to one key reference that probably can be avoided.

I guess my question is more about if on the security side I can trust the backend for SQL injection + if what am doing on the BB Builder is correct in terms of security to be sure the public user in the worst case scenario cannot have any scope on those part of columns.

Maybe am being just too risk adverse here and the schema table form option is enough secure, just wondering how the backend is handling security on a create Form

Let me know if it's clear

Thanks!

Lorenzo

**Comment by mjashanks** [comment, marked as answer]
> Maybe am being just too risk adverse here

no such thing! 

So. We have ensured that Budibase is protected against SQL injection while using either custom queries or the automatic table connections (i.e. "Fetch Tables").

So, there are 2 ways to read/write data from a table...

### 1. Automatic Tables - i.e. "Fetch Tables"

Whatever role you give access to these tables, you should consider them to have access to all of the columns in that table, regardless of 
- which fields are displayed in the frontend
- what filters are put on the front end
The reason for this is that the user with this role could always manually query the API under their login (thus bypassing any front end settings)

so - if "Basic" user has "Read" access - you should assume that they can read the entire table

### 2. Custom queries

The same rules apply, except
- They can only see columns that are included in the `SELECT` statement
- If a filter is applied at the SQL level (`WHERE ... `) then they cannot bypass that.
I.e. the are restricted by the query that you write.

Hope that answers the question!

**Reply by lollococce** [reply]
Oh ok, that was my intuition, thanks for confirming!

But if so I still have 3 questions

1. Let's say for the Form I wanna use a read SQL query to avoid using directly the table (image 1 below) (a Read SQL query with public access with a SELECT some_columns_only FROM table). I've tried but using a hierarchy Form > Field Group > Field I can't select the Query field in the field component (image 2 below) - is a Repeater or Data Provider needed or should I open an issue? This would be needed to show the options in the dropdowns of relationships (image 3 below)

<img width="244" alt="Screen Shot 2022-01-28 at 9 37 21 PM" src="https://user-images.githubusercontent.com/30072155/151617837-d3c89f23-f919-4aac-94ce-18f4f7aae0af.png">
<img width="236" alt="Screen Shot 2022-01-28 at 8 41 49 PM" src="https://user-images.githubusercontent.com/30072155/151618297-d589187b-e0e8-4b4b-96c8-fe277f13ec53.png">
<img width="1033" alt="Screen Shot 2022-01-28 at 9 40 27 PM" src="https://user-images.githubusercontent.com/30072155/151618244-5dc72d44-cc13-42e9-a234-976406acad32.png">


2. For the 'Save Row' part after compiling the form - is it so better to trust the 'Save Row' automation under the 'define action' of the Save button, or is it in this case (to hide some column names and access to insert) to execute a custom SQL INSERT query (given that the INSERT query defined is protected as you mentioned from SQL injection)?

3. Extra security question : For the form validation, defined under the 'Configure Validation' button, my question would be, is it this a frontend validation or a backend one? cause the frontend one can be bypassed and for example an email can be sent in theory without matching a proper email regex if the user is enough skilled to disable the JS on the front end

**Reply by mjashanks** [reply]
1. A few things 
- you have set the form to "Create", but your SELECT query suggests that you want to load data into the form - therefore, should it be an Update? 
- You will also need to have an `update` / `insert` statement, with a parameter for every field on the form.
- You would then use an `execute query` action to run the `insert`/`update` on save - mapping your form fields, to the query parameters.

2. In your case, use `Execute Query` plus a hand written custom SQL query. Otherwise you would need to give public write access to the whole table. Maybe that's ok - if so, `Save Row` will work

3. Configure validate is front-end only. At this moment, if you want to do backend validations, you could do so using the `Automations` section, which could run a check in a `Javascript` or `Filter` step, before saving the row in here. It's worth noting that automations run asynchronously, so you don't get feedback of the result to the client.

**Reply by lollococce** [reply]
All clear! Thanks so much for the answer!!
